providers/dynamodb/shedlock-provider-dynamodb/src/main/java/net/javacrumbs/shedlock/provider/dynamodb/DynamodbLockProvider.java

```diff
@@ -0,0 +1,306 @@
+     *                                                   .withEndpointConfiguration(
+     *                                                        new AwsClientBuilder.EndpointConfiguration(AMAZON_DYNAMODB_URL, REGION_US_WEST_2))
+     *                                                        .build();
+     *                                                   </pre>
+     * @param tableName              collection to store the locks
+     * @param minLockTime            minimum locking duration for a process. This is used to determine the required throughput
+     *                               for dynamo db. Use this as tuning parameter to optimize the cost and avoid any throttling
+     *                               by dynamo db. Higher minLockTime will reduce cost and also reduce throughput of dynamo db
+     * @param concurrentProcessCount number of concurrent nodes/processes running at a given time. This is used to
+     *                               determine the required throughput for dynamo db. Use this as tuning parameter
+     *                               to optimize the cost and avoid any throttling by dynamo db. Higher value
+     *                               will increase the cost and also increase throughput of dynamo db
+     */
+    public DynamodbLockProvider(AmazonDynamoDB amazonDynamoDB, String tableName, Duration minLockTime, long concurrentProcessCount) {
+        this(new DynamodbAccessor(amazonDynamoDB, tableName, minLockTime, concurrentProcessCount));
+    }
+
+    DynamodbLockProvider(DynamodbAccessor accessor) {
+        super(accessor);
+    }
+
+    public static class DynamodbAccessor extends AbstractStorageAccessor {
+        private boolean IS_TABLE_EXIST = false;
+        private final AmazonDynamoDB amazonDynamoDB;
+        private final String tableName;
+        private final String hostname;
+        private final Duration minLockTime;
+        private final long concurrentProcessCount;
+
+        DynamodbAccessor(AmazonDynamoDB amazonDynamoDB, String tableName, Duration minLockTime, long concurrentProcessCount) {
+            this.amazonDynamoDB = amazonDynamoDB;
+            this.tableName = tableName;
+            this.hostname = getHostname();
+            this.minLockTime = minLockTime;
+            this.concurrentProcessCount = concurrentProcessCount;
+            try {
+                createTableIfNotExists();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        @Override
+        public boolean insertRecord(LockConfiguration lockConfiguration) {
+
+            DynamoDB dynamoDB = new DynamoDB(amazonDynamoDB);
+            Table table = dynamoDB.getTable(tableName);
+            UpdateItemSpec spec = new UpdateItemSpec()
+                    .withPrimaryKey(NAME, lockConfiguration.getName())
+                    .withUpdateExpression(
+                            String.format("set %s=:%s, %s=:%s, %s=:%s",
```

Comment:
Can this whole string be a constant?

Reply:
One benefit on leaving it in the way it is would be a proximity to see the string template and their substitution values. Looking at 
String.format("set %s=:%s, %s=:%s, %s=:%s", LOCKED_BY, LOCKED_BY, LOCK_UNTIL, LOCK_UNTIL, LOCKED_AT, LOCKED_AT)) - the template/placeholder string is "set %s=:%s, %s=:%s, %s=:%s" and they are substituted by LOCKED_BY, LOCKED_BY, LOCK_UNTIL, LOCK_UNTIL, LOCKED_AT, LOCKED_AT. The reader can quickly visualize the update specification being formed here. 

If we make it a constant, then this wouldn't be very obvious as what is happening here. Imagine - String.format(INSERT_LOCK_UPDATE_SPEC, LOCKED_BY, LOCKED_BY, LOCK_UNTIL, LOCK_UNTIL, LOCKED_AT, LOCKED_AT)). This does not make it as clear. The reader needs to refer to the constant definition to grasp what might be going on here. 

A benefit of constant is usually re-use in multiple places so that changes are easy. We are a little localized here within this class. 

Thoughts?

Reply:
I would prefer the constant. For me the update expression itself is not of much interest, what I care about are the values of the variables. Moreover it just feels like a waste of processing resources to use String.format every time on something that's basically a constant. of course, it's just a feeling, the cost will be negligible compared to the network call that follows, but anyway.

Reply:
If we can have the following somewhere to be referred, should be enough for readability perspective. Creating constants for them by re using existing constants.

 String.format("set %s=:%s, %s=:%s, %s=:%s", LOCKED_BY, LOCKED_BY, LOCK_UNTIL, LOCK_UNTIL, LOCKED_AT, LOCKED_AT))